enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,754 @@
+
+* Has a Machine API equivalent AND has the authoritative annotation set with the `cluster.x-k8s.io` value
+* Does not have a Machine API equivalent
+
+By following the above rules, we should be able to ensure that the authoritative annotation is present in only one location and that the reconciliation of the objects is well defined based on the presence of this annotation.
+
+### Implementation Details/Notes/Constraints
+
+#### Synchronisation of resources
+
+To ensure that resources are synchronised, we will implement a new controller `machine-api-synchronisation-controller`.
+This controller will be responsible for translation of Machine API resources to Cluster API resources, and vice-versa.
+
+This controller will be responsible for creating Cluster API versions of Machine API resources, when they do not exist.
+It will observe the authoritative annotation described [above][authoritative-annotation] to synchronise between the two resources,
+overwriting anything in the non-authoritative resource that does not match with the authoritative resource.
+
+When errors occur during synchronisation, the controller will write conditions to the Machine API resource to report the failure.
+
+On successful sychronsiation, the controller will add an annotation to the non-authoritative resource to indicate the observed generation of the last successful synchronsiation.
+The annotation `machine.openshift.io/synchronised-generation` will be set with a value of the generation of the authoritative resource, onto the non-authoritative resource.
+This will allow users to keep track of, and ensure that the latest changes to the authoritative resource have been synchronised.
+
+When the `machine.openshift.io/synchronised-generation` is present on the authoritative resource, it will be cleared during synchronisation.
+
+[authoritative-annotation](#authority-of-resources-and-controller-reconciliations)
+
+##### MachineSet Synchronisation
+
+MachineSets in Machine API map to a MachineSet and InfrastructureTemplate in Cluster API.
+The synchronisation controller will create the Cluster API MachineSet with the same mame as the Machine API MachineSet to ensure a 1:1 mapping between these resources.
+The InfrastructureTemplate will be named based on the MachineSet name and a hash of the content within it [^6].
+InfrastructureTemplates are immutable, but MachineSet’s in Machine API are not, should the content of the MachineSet change,
+a new InfrastructureTemplate will be created and the old template will be removed when no longer required.
+
+The `providerSpec` forms the basis of the conversion to the InfrastructureTemplate in this case and is where errors may occur.
+Some values from the `providerSpec` in Machine API are moved to the InfrastructureCluster resource (a cluster wide infrastructure reference) and as such,
+if these values do not match that which is set on the InfrastructureCluster, we cannot convert losslessly.
```

Comment:
Not sure if this would actually work in OpenShift, but my understanding is that, in Azure, some information that exists on the Machine today, moves to the InfrastructureCluster. For example, the resource group to create the instance in, and this cannot be overridden locally to my knowledge. So this assumes all machines are in the same RG. If they aren't in a cluster today, we won't know how to reconcile this.